### ISSUE TITLE
[UX] Replace the classic way of creating select lists/checkboxes/radio buttons with Options Element.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1005]
Classic way:

![select_list_field-settings_classic](https://cloud.githubusercontent.com/assets/2423362/8144572/fde973c0-1227-11e5-8491-6236dd932be0.png)

[Options Element](https://www.drupal.org/project/options_element):

![select_list_field-settings_options_element](https://cloud.githubusercontent.com/assets/2423362/8144575/02f6cf16-1228-11e5-9189-62125c5d1fcb.png)

We should get the approval of the maintainer first though :stuck_out_tongue: 

---
Port of the D7 contrib module by @laryn: https://github.com/backdrop-contrib/options_element
~~PR by @serundeputy: https://github.com/backdrop/backdrop/pull/972~~
PR by @herbdool: https://github.com/backdrop/backdrop/pull/2204

**Comment by klonos** [comment]
@quicksketch: I saw a couple of issue in the d.org queue were you mentioned that you'd like to see this in core. Particularly interesting is the [Provide a reusable API to convert textarea fields to an options element](https://www.drupal.org/node/1980868) issue because if that happened, then we could use it in #1006.

**Comment by quicksketch** [comment]
Thanks @klonos. I like options element myself and we obviously have the ability to put it in core now. I think options element is an improvement on the UX, but for a long-time Drupal user I can't tell if it would be a welcome change or not. I'd like to hear other people's thoughts on the usefulness/intuitiveness of options element. This would be a great candidate for a usability study if we had the resources to do so.

**Comment by klonos** [comment]
> ...but for a long-time Drupal user I can't tell if it would be a welcome change or not.

I'm a long-time user and I welcome the change. @docwilmot's opinion coming up too...

**Comment by Graham-72** [comment]
I, too, would welcome the change.

**Comment by docwilmot** [comment]
I welcome this.

**Comment by serundeputy** [comment]
I installed the  [options_element] (https://www.drupal.org/project/options_element) on the D7 and I did not see the new/promised interface.

Is there configuration required?

**Comment by quicksketch** [comment]
It should be there. Make sure you've installed the latest stable (1.12) or used the 7.x-1.x branch. Apparently there's a master branch that's outdated.

**Comment by serundeputy** [comment]
yep; i was on `master` thanks!  :bow:

**Comment by serundeputy** [comment]
- there are no tests
  - the drag and drop crosshairs to rearrange the order of the options are not present
    - I'm assuming something with the javascript.
  - there is no config / there were no `variable_get`s or `variable_set`s 
    - so maybe no config is needed?

PR: https://github.com/backdrop/backdrop/pull/972

**Comment by serundeputy** [comment]
Issue #1005: fixing path to js for tabledrag.js and jquery.cookie.js.
updated PR.

**Comment by quicksketch** [comment]
Thanks for the PR @serundeputy! I think this element would make the most sense in options.module itself. Although some modules may wish to also use this element (e.g. Webform), I think it would probably be acceptable to depend upon the options.module for this functionality. Thoughts?

**Comment by serundeputy** [comment]
@quicksketch 
  * combine `options_element.module` with `options.module`
  * make `field/options/include/` put `options_element.inc` into it
  * mv `js` and `css` into `field/css` and `field/js` or should this be `field/options/css` ... ?
  * `add.png` and `delete.png` to `field/options/{images, assets}?

I don't have preference for where it is or with field/options, but if it is cleaner, better, more efficient to be there I'm glad to move it.  Let me know if the above steps seem right; and/or clarify the ones that are questions.

thanks,
~Geoff

**Comment by docwilmot** [comment]
aside, not quite on topic, but:
> add.png and delete.png 

I dont think those fit with the current admin theme aesthetic. Can we change them?

**Comment by quicksketch** [comment]
> make field/options/include/ put options_element.inc into it

Our current pattern per https://github.com/backdrop/backdrop-issues/issues/486 is that we leave .inc files in the root, until we establish an official decision in https://github.com/backdrop/backdrop-issues/issues/558. I'd say let's name it "options.element.inc".

> mv js and css into field/css and field/js or should this be field/options/css ... ?

These should be in `field/options/css`. We actually have an issue (and PR) to move all these submodules outside of the `field` module directory at https://github.com/backdrop/backdrop-issues/issues/557.

>  add.png and delete.png to `field/options/{images, assets}?

They should go in field/options/images/* yep. Also I think @docwilmot is right that these don't mesh with our current admin theme, but I think they work fine as generic elements. We should probably add Seven-specific styling in Seven's style.css.

**Comment by klonos** [comment]
I would like to "resurrect" this ...and target for 1.4.0 if possible?

@serundeputy if you're not busy with Real Life or anything more important for Backdrop core, if you get the chance to update the PR with the suggestions made, I'll make some time to test and review.

**Comment by serundeputy** [comment]
wait; there are things more important than Backdrop?  ... :bowtie:

**Comment by klonos** [comment]
:smile:

> or anything more important for Backdrop core

I meant any other, more important issues you're working on.

**Comment by laryn** [comment]
I ported [options_element](https://github.com/backdrop-contrib/options_element) in support of @jackaponte's [webform_civicrm](https://github.com/jackaponte/webform_civicrm/) work (which lists options_element as a dependency) but just came across this thread.

Options_element is not in core yet, is it?

**Comment by docwilmot** [comment]
Isn't yet. Feel free to continue where @serundeputy left off. :)

**Comment by herbdool** [comment]
Reviving this with [PR](https://github.com/backdrop/backdrop-issues/issues/1005). The one thing I haven't addressed was @docwilmot mentioning that the add/remove images don't match the aesthetic. Could that be a follow-up?

**Comment by laryn** [comment]
Took it for a quick spin and it looks good to me from a UI/usage perspective (noting that there may be a follow-up adjustment to the add/remove images).

Is there a way to disable the contrib `options_element` by default, too?

**Comment by docwilmot** [comment]
Nice work @herbdool 
- on creating a new field if you click manual entry the fieldset for Options settings remains, but then check/unchecking the 'Custom keys' checkbox has no effect. Maybe also hide the whole Options settings fieldset with JS if manual clicked?
- do we need the Option settings at all? would it simplify the UI (and JS) to just have key and value all the time, and simply tell the user "The value is optional: if a line contains only a key, it will also be used as the value."?
- the field add/edit screen is normally divided into sections for the particular content type being created/edited, and a section for global settings. BUT with this PR the "Default value" is now moved to the global section. Is that intentional? We cant do per-content-type default values with this? Or is it global values that are gone?
- simple niggle, when adding key/values, hitting tab moves from key>value>+>X>drag handle>checkbox>key. Can we make it only tab to key>value>nextkey>nextvalue?
- Not sure I like the "No default" link. I hadnt noticed the checkbox column for defaults and so was clicking that link wondering what its supposed to do. Would a "select/unselect all" checkbox in the tableheader be more intuitive? P.s. can a user actually have more than one default value in an option element? If not maybe the JS should enforce this.
- would also move "Add item" to where "no default" is, if we could.

Only UX review, havent checked code yet.

**Comment by laryn** [comment]
@herbdool I agree that would be good to do in addition but if there are folks who haven't upgraded the contrib module it would still be nice to neutralize it for them when this goes in core. Maybe that's just opening a can of worms though (e.g. if they also have webform_civicrm installed and not upgraded, which lists options_element as a dependency...)

**Comment by docwilmot** [comment]
Re 2, I'm also not sure, But keep in mind the description now in Backdrop is: "The label is optional: if a line contains a single string, it will be used as key and label." People would be already used to that. And the UI would be simpler. I think, not sure.
Re 4 unless this has to do with accessibility, then skipping would be surely better, no? You click "Add" 10 times to create 10 options, then you tab, type, tab, type. More convenient I think.
Re 5 agreed
Re 6 and 7 yes you can have multiple defaults youre right, so Cant move that.

**Comment by herbdool** [comment]
@docwilmot 

Re 2: Try it yourself without this PR. When creating the field it'll automatically duplicate the labels to be the values too. So even if you type in

> One
> Two
> Three

It'll be turned into

> One|One
> Two|Two
> Three|Three

So I think your suggestion is better as a follow-up since it would mean changing the current behaviour without the widget as well.

Re: 4: I think accessibility may be a concern but I don't know enough about it to say how it should be. Perhaps this can also be a follow-up.

**Comment by docwilmot** [comment]
@herbdool I wonder if we're misunderstanding each other here. I mean always show this:

![optionexp](https://user-images.githubusercontent.com/1042283/40700933-b6714846-63d3-11e8-92c3-36b8f1b14f1f.PNG)

Get rid of this functionality:

![optionnokey](https://user-images.githubusercontent.com/1042283/40700948-cc97a976-63d3-11e8-8b0d-e92dfdb80cef.PNG)

>So even if you type in
It'll be turned into

Exactly. So keeping the UI as in the first image is more transparent as it shows that there will always be a key, Backdrop will just add it for you if you dont type it yourself. THe JS custom keys checkbox is a bit of a lie as it suggests thats how things will be saved.

**Comment by herbdool** [comment]
@docwilmot I think I did understand you. May it's more "transparent" but it is also changing the current default behaviour which doesn't force people to enter keys. In your proposal how do you propose to encourage people to only enter values and have Backdrop automagically create keys as it currently is behind the scenes?

I think there's a good reason to hide it when initially adding since otherwise people would think they were required to add keys, when Backdrop isn't actually requiring them.

My feeling is that your proposal is best as a separate issue since it would probably result in a new default functionality from that which existed prior to this PR.

**Comment by docwilmot** [comment]
> @herbdool it is also changing the current default behaviour which doesn't force people to enter keys

I was not suggesting changing any behavior, just the UI. I propose this only to simplify things. If a user enters only a key we then have Backdrop automagically create a label from that key, just as it is now in D7/B1. Core currently says "The label is optional: if a line contains a single string, it will be used as key and label".

**Comment by herbdool** [comment]
@docwilmot I still think you're suggesting a change to current behaviour. You are suggesting that the keys be always visible. Even if technically the keys are still optional, to most people that will _appear_ as if the keys are required.

If we did not have this toggle, then I can imagine someone soon coming up with a use case for building one. If I may phrase this as a "user story" (to use agile project management parlance):

> "As a new user of Backdrop I would like the option to hide the option keys by default so I do not have to think about what keys to type in."

or from the perspective of an administrator of the site who has users making webforms:

> "As a site administrator I would like to hide the option keys by default so that it is easier for my users who are creating webforms."

In both cases, it doesn't matter if the keys are actually optional, the important thing is the end user not getting confused by having the key field visible.

(Which reminds me that I need to ensure this still works with Webform and Webform CiviCRM if this is merged into core.)

**Comment by klonos** [comment]
Here's my take on how the UI for this should look and work:

<img width="1174" alt="screen shot 2018-06-24 at 12 04 35 pm" src="https://user-images.githubusercontent.com/2423362/41815204-e5625564-77a6-11e8-868a-00a320035dc8.png">

So, basically...

- Remove the "Option settings" fieldset, as @docwilmot suggested.
- Move the "KEY" column after the "VALUE" column.
- Add a "required" red asterisk in the "VALUE" header.
- Make the key fields look and work more like the machine name thingy (noting that contrary to machine names, these should allow spaces and capital letters, in order to retain current functionality).
- Add zebra striping and borders to the table, and fix the `margin-left: -1em;` for the `tabledrag-processed` classes.
- Rename the "No defaults" link to "Clear default(s)".
- Move the "Manual entry"/"Normal entry" to the top-right of the table (same position we use for the "Show row weights"/"Hide row weights" toggle). I really find it terrible UX to have this toggle in the same place as the "Add item" link.
- Style the "Add item" link as a button and change its label to "Add another item", to match the button of fields with multiple/unlimited values.
- Change the current action icons/buttons to drop-buttons with "Add item after" and "Delete item" as actions, and give the column an "Operations" heading.

**Comment by klonos** [comment]
...the only other alternative that I can think of is to have all keys be automatically generated by default, have their fields be disabled, and offer a "Custom keys"/"Auto-generated keys" toggle:

<img width="1179" alt="screen shot 2018-06-24 at 12 26 51 pm" src="https://user-images.githubusercontent.com/2423362/41815308-0b021be4-77aa-11e8-83d0-0a5cdb983769.png">

<img width="1174" alt="screen shot 2018-06-24 at 12 27 38 pm" src="https://user-images.githubusercontent.com/2423362/41815310-0d471df0-77aa-11e8-8f54-14a81355ab87.png">

This will have the following benefits:

- does not change the order of key/value columns/fields (going with the machine name thingy would have)
- we can skip the tab index for disabled key fields when the toggle is set to auto-generate them 😉

**Comment by docwilmot** [comment]
Thanks @klonos but after all that I'm now leaning back towards leaving the UI as it is now in @herbdool's PR with only the label field showing, and an option to toggle the keys. Had a think after watching your mockups and re-reading the issue, and I think that is in fact simpler.

Except remove the"Options settings" fieldset and replacing with just a simple text toggle above the area, like the in the image just above this.

**Comment by klonos** [comment]
Another thought about something that makes sense from an end user point of view after reading your comment @docwilmot is that we should change "value" to "option label" or just "label". The thought being that the things the user is adding here are *options* (or *items*) to the radio/checkbox/select element.

Thoughts?

**Comment by docwilmot** [comment]
I think I like value probably better; people might wonder what theyre labeling. Or maybe just "item"

**Comment by klonos** [comment]
> ...people might wonder what theyre labeling.

You're absolutely right @docwilmot. Perhaps this table needs a label itself for clarity. Something like "Checkbox/radio options" or "Select list options" (depending on which widget the user has selected).

Related(?)... I am still not clear why we are using checkboxes instead of radios in the "Defaults" column. AFAIKT, the only way to have multiple defaults is if the widget is a select list. In the case of a set of checkboxes/radios, there can only be a single default. Right?

**Comment by herbdool** [comment]
A few things to keep in mind:

* This has got to work with Webform as well as core, since that's how the original module that this potentially replaces worked.
* During manual entry the language is "key|value" so both the types of entry need to be consistent. So this change is better as a follow-up issue so it can be changed across the board if people agree.
* Checkboxes and radios in the defaults column is determined by whether the field is single or unlimited. So that works as expected.
* The "Custom keys" checkbox has help text whereas the link toggle doesn't. I'm not sure if that matters or not.

My general feeling is that I prefer to keep the scope of this limited and create follow-up issues as much as possible. If I'm the one working on this, my time is pretty limited so this could take quite a long time to get into core if the scope of this keeps expanding. We still need to get the old option lists tests passing again, possibly create new tests, and I find tests can sometimes take quite some time to wrap my head around.

**Comment by klonos** [comment]
Totally aware/respectful of time limitations @herbdool ...my feedback had mainly to do with:

a) Consistency with the rest of the admin UI: using drop-buttons instead of +/X icons, table rows with zebra stripping, using link toggle to switch between a noob/advanced entry mode (thinking menu ordering by weight) etc.
b) simplifying the UI and the terminology used ("key"/"value" is database engineer lingo).

To minimise time waste, perhaps get some feedback from people like @wesruv and @jenlampton before starting to do further work?

**Comment by klonos** [comment]
@herbdool why did you mention this specifically?:

> This has got to work with Webform as well as core

...is there anything that does not currently work? ...or perhaps some of my recommendations are not inline with the Webform UI?

**Comment by herbdool** [comment]
Options Element was written to work with both content type fields and with Webform (and Webform CiviCRM by extension). Any suggested changes to the UX should see how they look in Webform.

Making changes to the UI for this feature is complicated and takes time to implement and add to tests. Even just moving the Custom keys checkbox to the top and turning it into a link has surprisingly taken me longer than I expected.

I like the idea of improving the UX, but my preference that most of these suggestions be made as follow-up issues for *after* the original feature is merged into core. Then we can make decisions based on the effort to implement and the importance. Even though some suggestions seem simple and quick, that hasn't turned out that way thus far. We have a real risk of bogging this down in an expanding scope and retreating horizon of when this is considered acceptable for core.

**Comment by herbdool** [comment]
There. I've taken it about as far as I can without reworking a bunch of stuff (which I've suggested above be added as follow-up issues for improving the UX). What's been tweaked:

* ensure tests pass. The `list (text)` field type works slightly differently with this option widget enabled. Namely, if no keys are specified it will assign them numerically rather than duplicate the value into the key. In my testing that's how the original contrib module worked too.
* `Manual entry` link is above the widget. This helps separate this toggle from the widget's internal working.
* Hide/show the custom key toggle depending on Manual entry toggle.
* Made the plus/minus icons greyscale to match Seven theme. (Can be overridden in a theme.)

**Comment by klonos** [comment]
The default column seems to now be entirely gone.

If we think this is good to go as is, then I do not mind filing follow-ups 👍

**Comment by herbdool** [comment]
@klonos the default column still exists. If you're looking at an existing field, it's under the Edit tab rather than settings.

**Comment by docwilmot** [comment]
@herbdool will this make it in? Whats missing?

**Comment by herbdool** [comment]
@docwilmot I think it's ready to be reviewed but I didn't update the labels on the issue. It's a MVP. Once it's in we'll need to update Webform.

**Comment by quicksketch** [comment]
I left a code review at https://github.com/backdrop/backdrop/pull/2204#pullrequestreview-151635803. Overall looks great! There's some work that needs to be done to update it a bit and fully integrate it into core instead of altering our own forms.

**Comment by herbdool** [comment]
`> My personal preference for the automated keys setting would be to display both the keys and values in the form when a field is being edited, but on creation only show the values (unless the keys were toggled on). That would solve both the problem of overwhelming people with too many options when backdrop can do that work for them, and also the problem of obscuring what's really going on.

From what I understand of your description, that's how options_elements works currently.

> some of these UI elements are unique to the contrib module, and should be replaced with core alternatives for consistency.

I'd like to note that options_elements is also used for webform so we need to make sure any UI changes work there as well.

**Comment by quicksketch** [comment]
From what I've seen, the current PR already fixes the problems with Options Element not looking like the rest of core. The colors have been desaturated and it matches everything else fairly well. We might want to increase the size of the plus/minus buttons a bit (possibly switch to SVGs)? The web is generally expected to be finger-friendly these days.

Here's the current sandbox:
![image](https://user-images.githubusercontent.com/100206/46170622-1b3f2f80-c253-11e8-8578-21cd56173c47.png)

Looks like @herbdool and @docwilmot have been having a lot of further discussion in the PR as well, check there for more updates. https://github.com/backdrop/backdrop/pull/2204

**Comment by herbdool** [comment]
Taking a cue from @quicksketch on increasing the size and switching to svg for the plus/minus, I've added replacement icons from the Material Design library https://material.io/tools/icons. It's licensed under Apache 2.0. I noticed that we already include Open Sans with the same license so this one should be okay. I can change it if we've got a different preferred library. I did see any. For now I included the two as separate SVG files but alternatively we could include an entire library as a font. http://google.github.io/material-design-icons/

I believe that the plus/minus is more user-friendly in this instance than drop-buttons. I've checked other similar applications, such as the checklists in Trello, and a number use similar interface.

By the way, one of the other suggestions above was to add striping to the table. Whenever I visualize how that would work on a dynamic table like this, I can't see how it would be usable or even look good. Because we can add new rows anywhere it means that any new row in the middle will cause all the rows following to switch striping. I think that would be really off-putting and distracting.

**Comment by herbdool** [comment]
@quicksketch I don't think I can completely avoid form alters. I still have `list_form_field_ui_field_edit_form_alter` so that the options "default widget" can be used instead of the field UI's default widget. I considered using a hook but since it needs to check the field instance to set variables in the field, I don't know if the hook can help.

**Comment by jenlampton** [comment]
There's no way that logic can go directly into `field_ui_field_edit_form()`?

**Comment by herbdool** [comment]
@jenlampton that might not be ideal since it would mean field_ui would then depend on the list and options modules and vice versa.

**Comment by jenlampton** [comment]
No place for an if module_exists()?

**Comment by herbdool** [comment]
I suppose but is that actually better than a form alter? At least with a form alter the logic stays in the module. I noticed there's already an alter for widgets in the list module already.

**Comment by herbdool** [comment]
This is ready for user testing again.

**Comment by herbdool** [comment]
@jenlampton 

1. I've changed value to label.
2. It's been moved and hidden along with the manual entry textarea.
3. I was having more trouble with this one. It has to be visible for regardless of the toggle so that makes it harder.

**Comment by herbdool** [comment]
@jenlampton 3) is now moved to just below the Add item link. I think it works better there otherwise it breaks up the UI for "Add item" and "No default".

**Comment by jenlampton** [comment]
Yes, I LOVE it!

**Comment by klonos** [comment]
I really love this too! ...shamelessly plugging #2304 which would be equally cool IMO.

**Comment by jenlampton** [comment]
Ui looks fantastic, all questions on my code review have been resolved. RTBC from me!

**Comment by quicksketch** [comment]
I gave this one more code review and it looks great! I've merged this into 1.x for 1.12.0. Thanks @herbdool, @docwilmot, @klonos, @jenlampton, @serundeputy, and @laryn!
